lib/ansible/plugins/connection/aws_ssm.py

```diff
@@ -0,0 +1,535 @@
+        self._poll_stdout = select.poll()
+        self._poll_stdout.register(self._stdout, select.POLLIN)
+
+        # Disable command echo and prompt.
+        self._prepare_terminal()
+
+        display.vvv(u"SSM CONNECTION ID: {0}".format(self._session_id), host=self.host)
+
+        return session
+
+    @_ssm_retry
+    def exec_command(self, cmd, in_data=None, sudoable=True):
+        ''' run a command on the ssm host '''
+
+        super(Connection, self).exec_command(cmd, in_data=in_data, sudoable=sudoable)
+
+        display.vvv(u"EXEC {0}".format(to_text(cmd)), host=self.host)
+
+        session = self._session
+
+        mark_start = "".join([random.choice(string.ascii_letters) for i in xrange(self.MARK_LENGTH)])
+        mark_end = "".join([random.choice(string.ascii_letters) for i in xrange(self.MARK_LENGTH)])
+
+        # Wrap command in markers accordingly for the shell used
+        cmd = self._wrap_command(cmd, sudoable, mark_start, mark_end)
+
+        self._flush_stderr(session)
+
+        # Handle the back-end throttling
+        for c in cmd:
+            session.stdin.write(c.encode('utf-8'))
+            if self.is_windows:
+                time.sleep(15 / 1000.0)
+
+        # Read stdout between the markers
+        stdout = ''
+        begin = False
+        stop_time = int(round(time.time())) + self.get_option('timeout')
+        while session.poll() is None:
+            remaining = stop_time - int(round(time.time()))
+            if remaining < 1:
+                self._timeout = True
+                display.vvvv(u"EXEC timeout stdout: {0}".format(to_text(stdout)), host=self.host)
+                raise AnsibleConnectionFailure("SSM exec_command timeout on host: %s"
+                                               % self.instance_id)
+            if self._poll_stdout.poll(1000):
+                line = self._filter_ansi(self._stdout.readline())
+                display.vvvv(u"EXEC stdout line: {0}".format(to_text(line)), host=self.host)
+            else:
+                display.vvvv(u"EXEC remaining: {0}".format(remaining), host=self.host)
+                continue
+
+            if mark_start in line:
+                begin = True
+                if not line.startswith(mark_start):
+                    stdout = ''
+                continue
+            if begin:
+                if mark_end in line:
+                    for line in stdout.splitlines():
+                        display.vvvvv(u"POST_PROCESS: {0}".format(to_text(line)), host=self.host)
+                    returncode, stdout = self._post_process(stdout)
+                    break
+                else:
+                    stdout = stdout + line
+
+        stderr = self._flush_stderr(session)
+
+        return (returncode, stdout, stderr)
+
+    def _prepare_terminal(self):
+        ''' perform any one-time terminal settings '''
+
+        if not self.is_windows:
+            cmd = "stty -echo\n" + "PS1=''\n"
+            if PY3:
```

Comment:
This shouldn't be needed, just do `to_bytes(cmd, errors='surrogate_or_strict')` without the PY3 conditional. Doing `""` on Python results in a byte string so `to_bytes()` will just be a no-op for that version. Also one style we try to follow in Ansible it to prefix any byte strings to `b_`. That was we can easily infer the type of string being used based on the variable name. If you have an explicit native string (bytes on 2, unicode on 3) then we prefix with `n_`. Otherwise a general text string has no prefix.

Reply:
Updated in commit https://github.com/ansible/ansible/commit/8a5f982ce77b02d747c2f1954c3e413a449211f8